File: src/javascript/src/domain/entities/book.ts
import { ArgumentNullError } from '../errors/argument_null_error'
import { Author } from './author'
import { Price } from './price'
import { Publisher } from './publisher'
import { Review } from './review'

export class Book {
    
    public authors!: Author[]
    public publisher!: Publisher
    public title!: string
    public pages?: number

    private _reviews: Review[]
    public get reviews() { return this._reviews }

    private _pricing: Price[]
    public get pricing() { return this._pricing }

    constructor(init: Partial<Book>) {
        if (!init.title) throw new ArgumentNullError('title')
        if (!init.publisher) throw new ArgumentNullError('publisher')
        if (!(init.authors?.length ?? 0)) throw new ArgumentNullError('authors')
        Object.assign(this, init)

        this._reviews = []
        this._pricing = []
    }

    setPrice(value: number, startingAt?: Date): void {
        this._pricing.push(new Price({
            value, 
            startingAt: startingAt ?? new Date()
        }))
    }

    getPriceAt(date: Date): Price | null {
        return this._pricing
            .sort((a, b) => a.startingAt >= b.startingAt ? 1 : -1)
            .find((price) => price.startingAt <= date) ?? null
    }

    placeReview(rating: number, note?: string): void {
        this._reviews.push(new Review({rating, note}))
    }

    getAverageRating(): number | null {
        if (this._reviews?.length ?? 0) return null
        let avg = 0
        this._reviews?.forEach((review) => avg += review.rating)
        return avg / this._reviews.length
    }
}

File: src/dotnet/HelloMutation.Domain/Entities/Book.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace HelloMutation.Domain.Entities
{
    public record Book
    {
        public Book(
            string title,
            IEnumerable<Author> authors,
            Publisher publisher,
            ushort? pages = null
        ) => (Title, Authors, Publisher, Pages) = (title, authors, publisher, pages);

        private IEnumerable<Author> _authors;
        public IEnumerable<Author> Authors
        {
            get => _authors;
            init => _authors = (value?.Any()).GetValueOrDefault() ? value : throw new ArgumentException("", nameof(Authors));
        }

        private Publisher _publisher;
        public Publisher Publisher
        {
            get => _publisher;
            init => _publisher = value ?? throw new ArgumentException("", nameof(Publisher));
        }

        private string _title;
        public string Title
        {
            get => _title;
            init => _title = !string.IsNullOrWhiteSpace(value) ? value : throw new ArgumentNullException(nameof(Title));
        }

        public ushort? Pages { get; }
        public IList<Review> Reviews { get; } = new List<Review>();
        public IList<Price> Pricing { get; } = new List<Price>();

        public void SetPrice(decimal value, DateTime? startingAt = null) => Pricing.Add(new Price(value, startingAt ?? DateTime.Now));

        public Price GetPriceAt(DateTime date) => Pricing
            .OrderByDescending(price => price.StartingAt)
            .FirstOrDefault(price => price.StartingAt <= date) ?? new Price(0m, DateTime.Now);

        public void PlaceReview(ushort rating, string note = null) => Reviews.Add(new Review(rating, note));

        public ushort? GetAverageRating() => Reviews.Any() ? (ushort?)Reviews.Average(review => review.Rating) : null;
    }
}
